src/isl.rs

```diff
@@ -0,0 +1,211 @@
+    /// represents an unnamed type definition reference
+    AnonymousType(IslType),
+}
+
+// TODO: add a check for nullable type reference
+impl IslTypeRef {
+    /// Tries to create an [IslTypeRef] from the given OwnedElement
+    pub fn parse_from_ion_element(value: &OwnedElement) -> IonSchemaResult<Self> {
+        match value.ion_type() {
+            IonType::Symbol => {
+                value.as_sym().unwrap()
+                    .text()
+                    .ok_or_else(|| {
+                        invalid_schema_error_raw(
+                            "a base or alias type reference symbol doesn't have text",
+                        )
+                    })
+                    .and_then(|type_reference| {
+                        let ion_type = match type_reference {
+                            "int" => IslTypeRef::IslCoreType(IonType::Integer),
+                            "float" => IslTypeRef::IslCoreType(IonType::Float),
+                            "decimal" => IslTypeRef::IslCoreType(IonType::Decimal),
+                            "timestamp" => IslTypeRef::IslCoreType(IonType::Timestamp),
+                            "string" => IslTypeRef::IslCoreType(IonType::String),
+                            "symbol" => IslTypeRef::IslCoreType(IonType::Symbol),
+                            "bool" => IslTypeRef::IslCoreType(IonType::Boolean),
+                            "blob" => IslTypeRef::IslCoreType(IonType::Blob),
+                            "clob" => IslTypeRef::IslCoreType(IonType::Clob),
+                            "sexp" => IslTypeRef::IslCoreType(IonType::SExpression),
+                            "list" => IslTypeRef::IslCoreType(IonType::List),
+                            "struct" => IslTypeRef::IslCoreType(IonType::Struct),
+                            // TODO: add a match for other core types like: lob, text, number, document, any
+                            _ => IslTypeRef::AliasType(type_reference.to_owned()),
+                        };
+                        Ok(ion_type)
+                    })
+            }
+            IonType::Struct =>
+                Ok(IslTypeRef::AnonymousType(value.as_struct().unwrap().try_into()?)),
+            _ => Err(invalid_schema_error_raw(
+                "type reference can either be a symbol(For base/alias type reference) or a struct (for anonymous type reference)",
+            )),
+        }
+    }
+
+    /// Resolves a type_reference into a [TypeId] using the type_store
+    pub fn resolve_type_reference(
+        type_reference: &IslTypeRef,
+        type_store: &SharedTypeStore,
+    ) -> IonSchemaResult<TypeId> {
+        match type_reference {
+            IslTypeRef::IslCoreType(ion_type) => {
+                // TODO: create CoreType struct for storing ISLCoreType type definition instead of Type
+                // inserts ISLCoreType as a Type into type_store
+                Ok(type_store.borrow_mut().add_named_type(
```

Comment:
TODO for the future: `TypeStore` should be responsible for calling `borrow_mut()` internally; making users call it is a leaked implementation detail.